Tutorial for writing a service with an asynchronous client node (#4903)
ros2/ros2_documentation: source/Tutorials/Intermediate/Async-Service/Async-Service-Main.rst

> .. Tutorials/Intermediate/Async-Service

Reviewer: why do we need this redirection? since this is new doc, can we remove this?

Author: Of course.

Author: Done

Writing a service with an asynchronous client node
==================================================

.. contents:: Table of Contents
   :depth: 2
   :local:

**Goal:** Understanding asynchronism in the context of services.

**Tutorial level:** Intermediate

**Time:** 20 minutes

Background
----------

For information on how to write a basic client/server service see :doc:`Writing a Simple Service and Client <../../Beginner-Client-Libraries/Writing-A-Simple-Cpp-Service-And-Client>`.

In this tutorial we will create a service server with a artificial delay and a service client that makes requests to the service server **inside a callback**.
We will use both to understand the concept of asynchronism and to evaluate the impact of timing in the design of services.

The delay in the server will make it easier to observe client activity; that activity is the difference between synchronous and asynchronous requests.

> A **synchronous** client is idle (i.e. **doing nothing**) after making a request.

Reviewer: i am not sure if this is the correct description. IMO there is no such thing like `sync client` or `async client`, it is more like how client wants to send the request either async or sync manner? maybe we can ignore this.

Author: I don't think async or sync concepts have a wide and standard definition for every possible context. But, well, I also think there is consensus about that a synchronous request freezes the execution in the client until the response is received.
I don't think "ignore it" is a good idea. People should know what they are doing.

Author: I don't think async or sync concepts have a wide and standard definition for every possible context. But, well, I also think there is consensus about that a synchronous request freezes the execution in the client until the response is received.
I don't think "ignore it" is a good idea. People should know what they are doing.

Once it receives the response it resumes execution with the statement immediately following the request.

Conversely, an **asynchronous** client can **continue running** after making the request, performing other tasks.
Later, it may either wait for the response or execute a predefined callback.
In this tutorial, the client will execute a callback when the response is received.

Consider the following diagram:

.. image:: images/sync-client-diagram.png
   :target: images/sync-client-diagram.png

>    :alt: A sequence diagram that show how a sync-client waits for the response

Reviewer: this is where `mermaid` diagram can be rendered instead of image insertion.

.. image:: images/async-client-diagram.png
   :target: images/async-client-diagram.png
   :alt: A sequence diagram that show how a async-client is not waiting but doing something else

In general, an asynchronous client continues running after making the request.
Afterward, when the response is required to proceed, the thread could call a waiting function to stop until the response is received.
In a ROS 2 client, however, the thread must be spinning to receive the response.
This means it has to continuously poll for incoming events (topics, timers, etc.), including the response event.
This behavior is depicted in the diagram: when ``Event A`` is received, its callback is executed (the client is *activated*).
Later, the response is received, and the client executes the corresponding callback for the response.

Since any service client in ROS 2 must be spinning to receive the server's response, all clients in ROS 2 are inherently asynchronous by design.
Service requests cannot block the thread, and although they cannot simply wait, they can wait while spinning.

.. attention::

>    **All** service **clients** in ROS 2 are **asynchronous**.

Reviewer: then why we are introducing `sync client` here above?

and this is true, there is no such API for sync request for the service. `rclpy` actually has sync one that is why we have https://docs.ros.org/en/rolling/How-To-Guides/Sync-Vs-Async.html.

now i think the explanation above can be moved to https://docs.ros.org/en/rolling/How-To-Guides/Sync-Vs-Async.html.

what do you think?

Author: To your first question: As I said before, people should know what they are doing and what they can or can not do (should do or should not do). Understanding async request could be explained by showing the difference with sync requests. That is even more important in this case because as far as I know, in ROS 1 service request were sync.

An additional comment in this matter: even if there is an interface for doing a sync request in python, I think it is only in apparence, as I said somewhere it is not possible to send a truely sync request because ROS 2 has to spin to receive the response => the program is not idle => it is not sync. 

To your second question: Well, we could. But regarding the sync api... Well, in my very personal and humble opinion that api should be avoided at all cost (Something that has  in the first two paragraphs of its description the words: "risks", "deadlock" and "pitfalls" should be avoided)... Even more I think the origin of all these "risks", "pitfalls" is trying to convert something that is essentially async into sync.   

That could be the subject of a very thoughfull discussion and this is not the appropiate forum.  Anyway, that is my reason for not contribuiting there.

Reviewer 2: > as I said somewhere it is not possible to send a truely sync request because ROS 2 has to spin to receive the response => the program is not idle => it is not sync.

This isn't really true.  It is definitely subtle, but if you spin in one thread, and do a `client.call` in another thread, then it can be (and often is) the case that both threads are idle until the response to the client comes in.  In that case, it truly is "synchronous" in the thread that is calling `client.call`.

That said, my personal opinion is that the rclpy `client.call` API was a mistake, and we never should have added it.  If you understand what you are doing, and you truly want to do what `client.call` does, the equivalent code is:

```
future = client.call_async()
rclpy.spin_until_future_complete(future)
```

Author: Thanks @clalancette for your comment. 
Agreed, definitely is subtle. I was not considering multi-threading. And if you think about these issues deeper, you might find this question interesting: why to launch a thread just to keep it on hold?
And agreed again, actually I would mark that api as "deprecated" to discourage its use. But that is another discussion.

      response->sum = result;
      RCLCPP_INFO_STREAM(this->get_logger(), "Response: " << result);
   }

   rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr service_;
   };

   int main(int argc, char ** argv)
   {
      rclcpp::init(argc, argv);
      auto node = std::make_shared<DelayedSumService>();
      rclcpp::spin(node);
      rclcpp::shutdown();
      return 0;
   }

Update ``CMakeLists.txt`` to build the executable: add the following lines to it (after finding packages):

.. code-block:: console

   add_executable(service_main main.cpp)
   ament_target_dependencies(service_main rclcpp example_interfaces)

   install(TARGETS service_main DESTINATION lib/${PROJECT_NAME})

Then install dependencies if you need:

.. code-block:: bash

   rosdep install -i --from-path src --rosdistro {DISTRO} -y

And build as usual:

.. code-block:: bash

   colcon build

1.2 Examine the server code
~~~~~~~~~~~~~~~~~~~~~~~~~~~

This node has a standard implementation with nothing special.

.. warning::

   This package is a service server with an arbitrary and artificial delay in responding to requests.

>    It should not used unless the delay is removed.

Reviewer: i do not really understand why this delayed service server is required to explain this? if the client waits on the request in the callback, that will be deadlock regardless the server implementation. why dose this delayed server needs to be introduced as example?

Author: Well, I was already sure about the relevance of this tutorial, but reading your comments I am even more sure. 

There is NOT such deadlock. An async client is NOT waiting, idle or hold, it is spinning normally. I am not sure if it is clear enough, it is said in this way: "It is non-blocking, meaning it returns almost immediately without stopping the execution of the thread." And it is also said that: "the best approach is to allow the callback to finish" and that is what it is done in the example. Of course, you can not get your response until that response is computed and received, but that is not a deadlock. A basic principle for a framework like ROS 2 is "when event A occurs Callback_A will be executed", here the event is "receiving the response", so an async client is based on "when response A is received, Callback_A will be executed". Mind that I say "will" and not "is" because Callback_A will be executed as soon as other callbacks are resolved (depends on the executor scheduler).

On the other hand, I am worried, apparently you have not understood the actual concepts I am trying to explain and I am not sure on how to handle that. Besides, I don't understand your problem with the delayed server, if I use a server example that makes one million of matrix inversions, will it be an acceptable example? 

Honestly, I will not be able to work in this matter much longer, so, please, tell me if changing the uml diagrams to mermaid will be enough because I will only do that if you are going to accept the PR.

Reviewer: @santiago-tapia thanks for being patient and iteration.

Ah sorry i was meaning the deadlock case but that is wrong since you used `async_send_request` in the topic callback.

what i mean here is, do we really need to have this `delayed service` to explain this tutorial in addition to topic callback? i would just use timer callback to send the async request to the already existed https://github.com/ros2/examples/tree/rolling/rclcpp/services/minimal_service that also explains the `async` example that you are trying to explain here? it would be much less code and maintenance cost effective, does that make sense? (this also goes to https://github.com/ros2/examples/pull/398)

> please, tell me if changing the uml diagrams to mermaid will be enough because I will only do that if you are going to accept the PR.

i cannot guarantee this, sorry.

Author: @fujitatomoya, thanks again for your comment, and I apology if I sound impatient yesterday. Let's keep positive. 

All right, I will change the diagram to mermaid. 

And now, I will try to explain why "delayed server" is necessary to explain async. 

First, I would like to mention possible use case for services: querying a data base, do some image processing, make some slam computation, etc. With respect to the computing time to produce a response they have in common two features: the computation time is not immediate and it is not predictable. So the first reason to make "delayed server" is including in the example an emulator of a non-trivial service server (at least it will not produce a response immediately). 

Second, the whole tutorial is about understanding async service calls and its differences from sync calls (sync calls in general, as I say, there is no such a thing in ROS 2 due to its design). Even if there might be some subtle details according to the context, I think we could agree that: 

1) An async call returns **immediately**. 
2) So It can not have the response as an actual returned result (that result is in the "future").
3) There must be some language / framework mechanism for retrieving the response in the "future". 
4) The thread keeps running immediately at the next sentence. 

All of these features are important, and should be understood. The role of "delayed server" is to introduce an observable time interval between the moment the async request is sent and the moment in which is received. And it should help in understanding the previous points because: 

a) The node is capable of processing another topic callback before it gets the response from the server (in single thread). That confirms 1) and 4). By the way, I might add a LOG just before ending the callback to make it more explicit. The delay here is absolutely necessary because we need "plenty" of time to produce an incoming message to the client node. This also illustrates why a deadlock is impossible, the node is not waiting or hold in any way (apart from "waiting" inside the spin method).
b) The node is also capable of sending two requests in a row and afterwards retrieving the two responses without problem. This explains why callbacks in 3) are superior in retrieving responses over other alternatives (for instance to waiting). The role of the delay here is to have "plenty" of time to send the second request without receiving the response to the first one. 

I think that's all. Please, just tell me if it is clear enough now. 

Best

Reviewer: > I apology if I sound impatient yesterday

no worries.

> the computation time is not immediate and it is not predictable.

those are what i meant. what do you mean by `immediate`? that all depends on application and use cases, i think.
and my point here is not predictable, i mean server behavior or implementation is technically irrelevant to what you are trying to explain here on client async behavior.
this is the reason that i still think we can keep this only with client async example by calling async request via timer callback.

> make "delayed server" is including in the example an emulator of a non-trivial service server

i think just adding sleep in the server process can be still trivial.

> All of these features are important, and should be understood.

agree, but again all these are client side behavior, not relevant to server implementation.

after all, i still think we would not want to maintain `delayed server` as example here.
i would recommend more fundamental and simplified tutorial how to handle the async request for the server via user callback of the entity such as timer.
i think that is easier to understand, reusable and  primitive for developers and users.

and i do understand you are trying to explain this with your expectation for the application, but that is my opinion.
we can keep this open for more comments to see either way we proceed.

.. redirect-from::

   Async-Service

>    Tutorials/Async-Service

Reviewer: this is whole new section, why we need this redirection?

it specifies one or more alternate URLs from which the page can be redirected. It's useful for maintaining backward compatibility when documentation pages are reorganized or renamed.

Writing a service with an asynchronous client node
==================================================

.. contents:: Table of Contents
   :depth: 2
   :local:

**Goal:** Understanding asynchronism in the context of services.

**Tutorial level:** Intermediate

**Time:** 20 minutes

Background
----------

For information on how to write a basic client/server service see 

> :doc:`checkout this tutorial <../../Beginner-Client-Libraries/Writing-A-Simple-Cpp-Service-And-Client>`.

Reviewer: ```suggestion
For information on how to write a basic client/server service see :doc:`Writing a Simple Service and Client <../../Beginner-Client-Libraries/Writing-A-Simple-Cpp-Service-And-Client>`.
```

> **All** service **clients** in ROS2 are **asynchronous**. In this tutorial, an example

Reviewer 3: ```suggestion
**All** service **clients** in ROS 2 are **asynchronous**. 
In this tutorial, an example
```

Reviewer 3: @santiago-tapia we usually do one sentence per line and ROS 2 is always stylized `ROS 2`.

Reviewer 3: I wouldn't assume all readers fully understand what is meant by "asynchronous." We get a lot of undergraduate students reading the docs who may not have reached that point in their education. It would help readability to quickly define your terms and then move on to discussing everything else.

of a service client is analyzed to provide insight into this fact, explaining
the reasons for its asynchronous nature and some of the timing-related
consequences.

> 

Reviewer 3: There's a saying in technical writing that you should, "tell people what you are going to do, tell them what to do, and then tell them what they did." 

It may help readability if you preface this tutorial with that you are going to do. Perhaps something like, "In this tutorials we are going to create A and B, and use it evaluate the impact of C on D."

Consider the following diagram: 

.. image:: images/sync-client-diagram.png
   :target: images/sync-client-diagram.png
   :alt: A sequence diagram that show how a sync-client waits for the response

It shows a **synchronous** client. The client makes a request and waits for

> the response, meaning its thread is not running but blocked until the

Reviewer 3: ```suggestion
the response, meaning its thread is blocked (i.e. not running) until the
```

response is returned.

> In contrast, the following diagram illustrates an asynchronous ROS2 client:

Reviewer 3: ```suggestion
In contrast, the following diagram illustrates an asynchronous ROS 2 client:
```

.. In the definition diagram there is an invisible interaction, in white color, otherwise the activation bar could not be deactivated.

.. image:: images/async-client-diagram.png
   :target: images/async-client-diagram.png
   :alt: A sequence diagram that show how a async-client is not waiting but doing something else

In general, an asynchronous client continues running after making the request.

> Of course, it could be stopped by invoking a waiting function or another
> blocking function, but this diagram shows a ROS2 client, and in ROS2, to receive

Reviewer 3: ```suggestion
Of course, it could be stopped by invoking a waiting function or another
blocking function, but this diagram shows a ROS 2 client, and in ROS 2, to receive
```
ll 50-53 is a bit of a run-on sentence. I suggest breaking this up into two sentences.

the response, the thread must be spinning, meaning it has to poll for incoming
events (topics, timers, etc.), including the response event. This behavior is
depicted in the diagram: when ``Event A`` is received, its callback is executed
(the client is *activated*), and afterward, the response is received, and the
client executes the callback for the response.

> Since any service client in ROS2 needs to be spinning to receive the server's

Reviewer 3: ```suggestion
Since a service client in ROS 2 needs to be spinning to receive the server's
```

> response, all clients in ROS2 are asynchronous by design (they cannot simply

Reviewer 3: ```suggestion
response, all clients in ROS 2 are asynchronous by design (they cannot simply
```

wait; they can only wait while spinning).

Thus, if a node is already spinning and has requested a service inside a
callback (any callback), the best approach is to let the callback finish.
You can process the response later in another callback in the future. This
approach might seem unintuitive because the code to be executed after
receiving the response is not written immediately after making the request.
However, you will get used to it; you just need to write or locate that
piece of code elsewhere.

This tutorial demonstrates how to write an asynchronous client that
operates as shown in the diagram.

Prerequisites
-------------

In beginner tutorials, you learned how to :doc:`create a workspace <../../Beginner-Client-Libraries/Creating-A-Workspace/Creating-A-Workspace>` 
and :doc:`create a package <../../Beginner-Client-Libraries/Creating-Your-First-ROS2-Package>`.

In :doc:`Writing a Simple cpp Service and Client <../../Beginner-Client-Libraries/Writing-A-Simple-Cpp-Service-And-Client>` you
learned how to define a custom service that adds two ints, we will use that service again. 

> The source code in this tutorial is at `rclcpp examples <https://github.com/ros2/examples/tree/{REPOS_FILE_BRANCH}/rclcpp/services>`,

Reviewer 3: ```suggestion
The source code for this tutorial is located at `rclcpp examples <https://github.com/ros2/examples/tree/{REPOS_FILE_BRANCH}/rclcpp/services>`.
```

> you might get the code there directly. 

Reviewer 3: ```suggestion

```

   }

   rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr service_;
   };

   int main(int argc, char ** argv)
   {
      rclcpp::init(argc, argv);
      auto node = std::make_shared<DelayedSumService>();
      rclcpp::spin(node);
      rclcpp::shutdown();
      return 0;
   }

Update ``CMakeLists.txt`` to build the executable: add the following
lines to it (after finding packages):

.. code-block:: console

   add_executable(service_main main.cpp)
   ament_target_dependencies(service_main rclcpp example_interfaces)

   install(TARGETS service_main DESTINATION lib/${PROJECT_NAME})

Then install dependencies if you need: 

.. code-block:: bash

   rosdep install -i --from-path src --rosdistro {DISTRO} -y

And build as usual:

.. code-block:: bash

   colcon build

1.2 Examine the server code
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. warning::

   This package is NOT a real service server example but rather an
   instrument to experiment with and understand the consequences of
   timing in services. It includes an artificial and unnecessary
   delay in responding to requests. Nevertheless, it could
   serve as an example if the delay is removed.

> Actually, there are no particularly relevant elements here. The main

Reviewer 3: I think you could simplify this quite a bit. Just say that you've created a server with an arbitrary and artificial delay..

      }
   }

This callback checks the topic value and, if it is greater than or equal to
zero, prepares a request to the service using the new topic value and 100 as
arguments, and then sends the request.

Key points about ``async_send_request`` are:

* It is called inside a callback, meaning it is executed in the thread that
  is spinning the node.

* It is non-blocking, meaning it returns almost immediately without stopping
  the execution of the thread.

* It accepts a callback as an argument, ``handle_service_response``, which is 
  where the code will *jump* when the response is received.

* There are no additional statements after the call to ``async_send_request`` 
  in ``topic_callback``, so execution will exit this callback and return
  to the spinning method.

* Keep in mind that the node must be spinning to receive the server response.

* The ``future_result`` object can be ignored since the response will be
  handled in ``handle_service_response`` using the argument. However, it can
  also be used to track the *state* of the request if necessary.

The second callback is for receiving the server response. Note that,

> being a callback, it will be executed at the spinning thread. The code is

Reviewer 3: "at the spinning thread" 

This phrasing is a bit unclear to me. Perhaps something like, "Note that since this function uses a callback, it will be executed promptly by the node's main thread while still allowing the node to accomplish other tasks."

quite simple:

.. code-block:: C++

  void handle_service_response(
    rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFuture future)
  {
    auto response = future.get();
    RCLCPP_INFO(this->get_logger(), "Response: %ld", response->sum);

    // Publish response at output topic
    auto result_msg = std_msgs::msg::Int32();
    result_msg.data = response->sum;
    publisher_->publish(result_msg);
  }

The response is provided in the parameter future. It is retrieved in the first
line and logged. Afterward, the response can be processed as needed. In this
example, it is simply published to a topic.

.. note::
   

>    Compared to the code of a hypothetical synchronous client, the key

Reviewer 3: This block, 412-416 could use a bit of revision. It might help to state plainly what the code does, and then contrast with the alternative version.

   difference lies in where the code to be executed *after* obtaining
   the result is placed. In a synchronous call, it is written directly
   *after* the line that makes the request. In an asynchronous call, however,
   it is located in **another callback**.

Installing the examples directly
---------------------------------

You might get the packages directly from code sources (clone the git
repository in a workspace and colcon build them) or if you
are using Ubuntu and you follow the `installation instructions <https://docs.ros.org/en/{REPOS_FILE_BRANCH}/Installation.html>`,
you can install them using apt for your ROS 2 distro:

.. code-block:: bash

   sudo apt install ros-{REPOS_FILE_BRANCH}-examples_rclcpp_async_recv_cb_client ros-{REPOS_FILE_BRANCH}-examples_rclcpp_delayed_service

Study how client and server interact
------------------------------------

Whether you write the package yourself or directly install the example, this
section provides cases of study to illustrate how the client and server
interact and the impact of execution timing on their interaction.

Discover available components
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> To see what packages that contains *examples_* are registered and available

Reviewer 3: ```suggestion
To see what packages contain the term *examples_* are available
```

in your workspace, execute the following command in a terminal:

.. code-block:: bash

   ros2 pkg list | grep examples_

> The terminal will show a list of packages from ros2_examples, actually,

Reviewer 3: This could be shortened. Try to keep it simple for the new users. 😉
